students/YingGuo/session04/mailroom2.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
+
+#session04, mailroom2
+
+from collections import defaultdict
+data = defaultdict(list)
+Sample = {"Tom":[100,200], "Tim":[3000], "Ken": [888,888,888], "Ted":[8000,5000,500], "Jon":[1000,6000]}
+for key in Sample.keys():
+    data[key] = Sample[key]
+
+#Send a thank you function 
+def thank_you(data):
+	while True:
+		user_input = input("please enter the name of donor. You can enter list to see the exisitng list of donors. Or enter quit to exist current function")
+		if user_input.lower() == "list":
+			print(data.keys())
+
+		elif user_input.lower() == "quit":
+			break
+		
+		else:
+			amount1 = int(input("how much would you like to donat?"))
+			data[user_input.capitalize()] += [amount1]
+			print("Thank you {} for your donation {}!".format(user_input.capitalize(), amount1))
+			return data
+			break
+
+
+#Creat a report function
+def report(data):
+	report_list = list()
+	for x in data.keys():
+		donation_sum = sum(data[x])
+		donation_num = len(data[x])
+		donation_average = donation_sum/donation_num
+		report_list.append((x, donation_sum, donation_num, donation_average))
+	
+	def custom_sorting(tuple):
+		return tuple[1]
+
+	report_list.sort(key = custom_sorting, reverse = True)
+
+	string_title = "{:<12}{} {:<12}{} {:<12}{} {:<12}".format("Donor Name","|","Total Given","|","Num Gifts","|","Average Gift")
+	print(string_title)
+	print("-"*len(string_title))
+
+	for x in report_list:
+		print("{:<12} ${:>12,.2f}  {:>12} ${:>12,.2f}".format(*x))
+
+def all_letter(data):
+    """Try to use a dict and the .format() method to produce the letter as one big template, the write into a txt file named with donor's name"""
+    for name in data.keys():
+        with open("{}".format(name),"w") as f:
```

Comment:
Good file writing code. Just wanted to mention that you might want to save these letter files in a subdirectory, not at the same directory... Besides, `"{}".format(name)` is redundant--it's just the same as `name`. You might have wanted to say something like `"letters/{}.txt".format(name)`? You'll need to create your subdirectory `letters` first, though...